docs/features/component-versioning.md

```diff
@@ -0,0 +1,88 @@
+
+The exact requirement list is TBD, but would likely include language version and specific component features we introduce going forward. These features could be dynamically calculated by the compiler: for instance, whilst still in preview a feature might only be enabled if the selected language version is also preview.
+
+While we'd allow authors to key on language version, we wouldn't be changing any of the existing functionality in the compiler to use these requirements: it would purely be for making decisions about whether to load the component or not.
+
+Features would also be able to represent super-sets of other features, for instance `LanguageVersion9` would imply `LanguageVersion1..8`. Components without the attribute would always be unconditionally loaded, with the same behavior as today.
+
+### Selective disabling
+
+A second use case for this feature would be to selective _disable_ the loading of certain components.
+
+Consider the migration of an `ISourceGenerator` to `IIncrementalGenerator`: the author may wish to create a new `IIncrementalGenerator`, but still fall back to the original `ISourceGenerator` in older compilers. The above proposal allows the user to *enable* the incremental generator when supported, but conversely they also need a mechanism to *disable* the `ISourceGenerator`.
+
+The `[RoslynFeature]` attribute would support a second, optional parameter that would cause the component to **not** load when the given feature is present.
+
+`[RoslynFeature("", disableWhen: "IncrementalGenerator")]`
+
+This would allow the author to provide a pair of components that enable/disable depending on features, allowing a smoother upgrade path.
+
+### Disable Warning
+
+In the above case, the author would not want the compiler to warn about either generator not loading, because a fallback is in place in both cases. The attribute would have a third optional parameter that would suppress the production of a warning:
+
+`[RoslynFeature("", warnOnNotLoaded: false)]`
+
+## Alternative Approaches
+
+There are a couple of obvious alternatives to the above proposal:
+
+### Version numbers
+
+Instead of being feature based, we could instead allow the author to specify the Minimum (and/or maximum) roslyn version that is required for the compiler.
+
+This has two downsides: it requires the author to explicitly know the version in which features are introduced; more importantly it doesn't allow us to move features into preview. 
+
+Imagine a feature that is introduced into roslyn version 3.8, but under a preview flag. How would a component specify that it should be loaded only when the specified feature is actually enabled?
```

Comment:
By allowing the generator to say "I depend on version 3.8, with preview features". In general, this feels like it's going to be a rehash of the questions we're going through with preview features in the language and runtime right now: https://github.com/dotnet/designs/pull/192. I strongly believe that we should just make it a version number and a flag for preview API features. We have all sorts of small API additions that would be problematic with a specific feature flag: would we need separate flags for each of these new APIs, or would we just make a misc flag?